scripts/functions-deploy-tests/tests.ts

```diff
@@ -0,0 +1,328 @@
+import * as path from "node:path";
+import * as fs from "node:fs/promises";
+
+import { expect } from "chai";
+import * as functions from "firebase-functions";
+import * as functionsv2 from "firebase-functions/v2";
+
+import * as cli from "./cli";
+import { Endpoint } from "../../src/deploy/functions/backend";
+import { assertExhaustive } from "../../src/functional";
+
+const FIREBASE_PROJECT = process.env.GCLOUD_PROJECT || "";
+const FUNCTIONS_DIR = path.join(__dirname, "functions");
+const FNS_COUNT = 12;
+
+function genRandomId(n = 10): string {
+  const charset = "abcdefghijklmnopqrstuvwxyz";
+  let id = "";
+  for (let i = 0; i < n; i++) {
+    id += charset.charAt(Math.floor(Math.random() * charset.length));
+  }
+  return id;
+}
+
+interface Opts {
+  v1Opts: functions.RuntimeOptions;
+  v2Opts: functionsv2.GlobalOptions;
+
+  v1TqOpts: functions.tasks.TaskQueueOptions;
+  v2TqOpts: functionsv2.tasks.TaskQueueOptions;
+
+  v1IdpOpts: functions.auth.UserOptions;
+  v2IdpOpts: functionsv2.identity.BlockingOptions;
+}
+
+async function setOpts(opts: Opts) {
+  let stmt = "";
+  for (const [name, opt] of Object.entries(opts)) {
+    if (opt) {
+      stmt += `export const ${name} = ${JSON.stringify(opt)};\n`;
+    }
+  }
+  await fs.writeFile(path.join(FUNCTIONS_DIR, "options.js"), stmt);
+}
+
+async function listFns(stripId = "dvtuqrxfjr"): Promise<Record<string, Endpoint>> {
+  const result = await cli.exec("functions:list", FIREBASE_PROJECT, ["--json"], __dirname, false);
+  const output = JSON.parse(result.stdout);
+
+  const eps: Record<string, Endpoint> = {};
+  for (const ep of output.result as Endpoint[]) {
+    const id = ep.id.replace(`${stripId}-`, "");
+    if (ep.id === id) {
+      // This endpoint is from another run. Ignore.
+      continue;
+    }
+    eps[id] = ep;
+  }
+  return eps;
+}
+
+describe("firebase deploy", function (this) {
+  this.timeout(1000_000);
+
+  const RUN_ID = genRandomId();
```

Comment:
Consider checking for (and providing) a relatable ID like the GH action number. Or, make sure to log this value so that if debugging needs to happen, it's easy to find (and then reference in the console)

Reply:
Opted for logging since I anticipate we run this test outside of Github Actions setting.